janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java

```diff
@@ -862,36 +871,38 @@ public void restore(Map<String,Map<String, List<IndexEntry>>> documents, KeyInfo
     }
 
     @Override
-    public List<String> query(IndexQuery query, KeyInformation.IndexRetriever informations, BaseTransaction tx) throws BackendException {
-        ElasticSearchRequest sr = new ElasticSearchRequest();
+    public Stream<String> query(IndexQuery query, KeyInformation.IndexRetriever informations, BaseTransaction tx) throws BackendException {
+        final ElasticSearchRequest sr = new ElasticSearchRequest();
         final Map<String,Object> esQuery = getFilter(query.getCondition(), informations.get(query.getStore()));
         sr.setQuery(compat.prepareQuery(esQuery));
         if (!query.getOrder().isEmpty()) {
-            List<IndexQuery.OrderEntry> orders = query.getOrder();
+            final List<IndexQuery.OrderEntry> orders = query.getOrder();
             for (int i = 0; i < orders.size(); i++) {
-                IndexQuery.OrderEntry orderEntry = orders.get(i);
-                String order = orderEntry.getOrder().name();
-                KeyInformation information = informations.get(query.getStore()).get(orders.get(i).getKey());
-                Mapping mapping = Mapping.getMapping(information);
-                Class<?> datatype = orderEntry.getDatatype();
+                final IndexQuery.OrderEntry orderEntry = orders.get(i);
+                final String order = orderEntry.getOrder().name();
+                final KeyInformation information = informations.get(query.getStore()).get(orders.get(i).getKey());
+                final Mapping mapping = Mapping.getMapping(information);
+                final Class<?> datatype = orderEntry.getDatatype();
                 sr.addSort(orders.get(i).getKey(), order.toLowerCase(), convertToEsDataType(datatype, mapping));
             }
         }
         sr.setFrom(0);
-        if (query.hasLimit()) sr.setSize(query.getLimit());
-        else sr.setSize(maxResultsSize);
+        if (query.hasLimit()) {
+            sr.setSize(Math.min(query.getLimit(), batchSize));
+        } else {
+            sr.setSize(batchSize);
+        }
 
         ElasticSearchResponse response;
         try {
-            response = client.search(getIndexStoreName(query.getStore()), useMultitypeIndex ? query.getStore() : null, compat.createRequestBody(sr, NULL_PARAMETERS));
-        } catch (IOException e) {
+            response = client.search(getIndexStoreName(query.getStore()), useMultitypeIndex ? query.getStore() : null, compat.createRequestBody(sr, NULL_PARAMETERS), sr.getSize() >= batchSize);
```

Comment:
Define index and type on separate line to get this line length a little more manageable.